### PR TITLE
fix #717 Adding user class

### LINKED ISSUE #717: Create a User class
**What would you like to be added**:
We currently have a feed class in `feed.js`, a post class in `post.js`, but we need one more, a user class in `user.js` to use the functions we are going to add to `storage.js` in #706 . Good references would be the other existing classes we have currently to give you an idea.

All code that interact directly with Redis will be in `storage.js`, any code that uses the functions we created in `storage.js` will be used by either `feed.js`, `post.js` or the yet to be created `user.js`

**Why would you like this to be added**:
We need this to allow users to add feeds

### PR DESCRIPTION
**Pull request by Grommers00**
Co-authored-by: David Humphrey <david.humphrey@senecacollege.ca>
Co-authored-by: Josue <josue.quilon-barrios@senecacollege.ca>

## Issue This PR Addresses

Automatically close the issue when this PR is merged
    USAGE: Fixes #717  

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

This will add the user class for ease of access later on when we get myFeeds working. This is just a draft PR and not ready to be added yet. 

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: src/backend/data/feed.js [hunk 1]
```diff
@@ -115,6 +116,14 @@ class Feed {
     const id = urlToId(url);
     return this.byId(id);
   }
+
+  /**
+   * Returns all the feeds
+   * Returns a Promise<Feeds>
+   */
+  static all() {
+    return getFeeds();
```
Comment (humphd) [inline src/backend/data/feed.js:125]: We will need to file a follow-up to fix all our other calls to `getFeeds()` and replace with this.

I think that `Feeds.all()` should be returning Feed objects.  So this should be something like:

```js
static async all() {
  const ids = await getFeeds();
  return Promise.all(ids.map(Feed.byId));
}
```

File: src/backend/data/user.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    return {
+      name: this.name,
+      email: this.email,
+      id: this.id,
+    };
+  }
+
+  async owns(feed) {
+    if (typeof feed.id === 'number') {
+      return this.owns(await Feed.byId(feed));
```
Comment (humphd) [inline src/backend/data/user.js:20]: With my change above, you'll have a Feed object already here.

File: src/backend/data/user.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+const Feed = require('./feed.js');
+
+class User {
+  constructor(name, email, id) {
+    this.name = name;
+    this.email = email;
+    this.id = id;
+  }
```
Comment (Silvyre) [inline src/backend/data/user.js:8]: Our user schema should look like this:
https://github.com/Seneca-CDOT/telescope/blob/b48c614d0e84b4e5bc8036938b7104821861eb94/src/backend/web/__mocks__/authentication.js#L14-L18

- We need an `isAdmin` data member.
- We should not have `id` as a constructor parameter. Instead, we should initialize user IDs by hashing the given email.

File: src/backend/data/user.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+  toJSON() {
+    return {
+      name: this.name,
+      email: this.email,
+      id: this.id,
+    };
+  }
```
Comment (Silvyre) [inline src/backend/data/user.js:16]: Why is this necessary?

File: src/backend/web/authentication.js [hunk 5]
```diff
@@ -70,7 +71,7 @@ function init() {
     done(null, user);
   });
   passport.deserializeUser(function (user, done) {
-    done(null, user);
+    done(null, new User(user.name, user.email, user.id));
```
Comment (Silvyre) [inline src/backend/web/authentication.js:74]: Once `user.id` has been refactored as above (to be initialized using a hashed email address), remove `user.id` from this call to construct a new User.

File: src/backend/data/user.js (hunk 3 above)
Comment (humphd) [reply src/backend/data/user.js:8]: The layer that they are working at is lower down than what you're describing.  The `isAdmin` stuff happens higher up.  Also, they need to accept the `id` as it comes from passport.

File: src/backend/data/user.js (hunk 4 above)
Comment (humphd) [reply src/backend/data/user.js:16]: This will override how the user object gets stored as JSON (e.g., JSON.stringify() will use it) when express writes this object to the session.

File: src/backend/web/authentication.js (hunk 5 above)
Comment (humphd) [reply src/backend/web/authentication.js:74]: They have to keep it in, since we're pulling it off the session for use in later code.
